tensorflow/lite/kernels/stablehlo_sort.cc

```diff
@@ -0,0 +1,267 @@
+#include "tensorflow/lite/kernels/internal/compatibility.h"
+#include "tensorflow/lite/kernels/internal/tensor.h"
+#include "tensorflow/lite/kernels/internal/types.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/kernels/tensor_slice_util.h"
+
+namespace tflite {
+namespace ops {
+namespace builtin {
+namespace stablehlo_sort {
+
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+
+std::vector<int> FlatToTensorIndex(int flat_index, const int dims,
+                                   const RuntimeShape& shape) {
+  int div = 1;
+  std ::vector<int> index(dims);
+  for (int i = dims - 1; i >= 0; --i) {
+    index[i] = (flat_index / div) % shape.Dims(i);
+    div *= shape.Dims(i);
+  }
+  return index;
+}
+
+template <typename DataType>
+TfLiteStatus EvalImpl(TfLiteContext* context, TfLiteNode* node,
+                      const std::vector<const TfLiteTensor*>& operands,
+                      const int dimension, const bool is_stable,
+                      std::vector<TfLiteTensor*>& results) {
+  const TfLiteStablehloSortParams* data =
+      reinterpret_cast<TfLiteStablehloSortParams*>(node->builtin_data);
+  Subgraph* this_subgraph = reinterpret_cast<Subgraph*>(context->impl_);
+  auto* subgraphs = this_subgraph->GetSubgraphs();
+  Subgraph& comparator_subgraph =
+      *(*subgraphs)[data->comparator_subgraph_index];
+  const int num_inputs = node->inputs->size;
+  const int num_outputs = node->outputs->size;
+
+  const TfLiteTensor* operand = operands[0];
+  const int rank = NumDimensions(operand);
+  const int operand_size = NumElements(operand);
+
+  std::vector<int> operand_index(rank);
+  std::vector<int> output_index(rank);
+  std::vector<DataType> args(operands.size() * 2);
+
+  const int dim_size = GetTensorShape(operand).Dims(dimension);
+  std::vector<int> indices(dim_size);
+  for (int k = 0; k < operand_size; ++k) {
+    auto result_it = FlatToTensorIndex(k, rank, GetTensorShape(operand));
+    if (result_it[dimension] != 0) continue;
+    std::iota(indices.begin(), indices.end(), 0);
+
+    auto comparator_wrapper = [&](int64_t lhs_handle,
+                                  int64_t rhs_handle) -> bool {
+      auto lhs_index = result_it;
+      auto rhs_index = result_it;
+      lhs_index[dimension] = lhs_handle;
+      rhs_index[dimension] = rhs_handle;
+      const DataType* operand_data = GetTensorData<DataType>(operand);
+      int flat_lhs_index = TensorIndexToFlat(lhs_index.data(), lhs_index.size(),
+                                             GetTensorShape(operand));
+      int flat_rhs_index = TensorIndexToFlat(rhs_index.data(), rhs_index.size(),
+                                             GetTensorShape(operand));
+      for (int x = 0; x < operands.size(); ++x) {
+        args[2 * x] = GetTensorData<DataType>(operands[x])[flat_lhs_index];
+        args[(2 * x) + 1] =
+            GetTensorData<DataType>(operands[x])[flat_rhs_index];
+      }
```

Comment:
It seems as though we write the values to this intermediate std::vector, and then write it to the subgraph input. Can we just write it directly to the subgraph input and skip the intermediate vector?